clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java

```diff
@@ -181,33 +243,461 @@ public void updateState(ConsumerGroupHeartbeatResponseData response) {
+
+            reconciliationResult.whenComplete((result, error) -> {
+                if (error != null) {
+                    log.error("Reconciliation failed. ", error);
+                } else {
+                    if (state == MemberState.RECONCILING) {
+
+                        // Make assignment effective on the broker by transitioning to send acknowledge.
+                        transitionTo(MemberState.ACKNOWLEDGING_RECONCILED_ASSIGNMENT);
+
+                        // Make assignment effective on the member group manager
+                        this.currentAssignment = assignedPartitions;
+                        this.targetAssignment = Optional.empty();
+
+                    } else {
+                        log.debug("New assignment processing completed but the member already " +
+                                "transitioned out of the reconciliation state into {}. Interrupting " +
+                                "reconciliation as it's not relevant anymore,", state);
+                        // TODO: double check if subscription state changes needed. This is expected to be
+                        //  the case where the member got fenced, failed or unsubscribed while the
+                        //  reconciliation was in process. Transitions to those states update the
+                        //  subscription state accordingly so it shouldn't be necessary to make any changes
+                        //  to the subscription state at this point.
+                    }
+                }
+            });
+        });
+    }
+
+    /**
+     * Build set of TopicPartition (topic name and partition id) from the assignment received
+     * from the broker (topic IDs and list of partitions). For each topic ID this will attempt to
+     * find the topic name in the metadata. If a topic ID is not found, this will request a
+     * metadata update, and the reconciliation will resume the topic metadata is received.
+     *
+     * @param assignment Assignment received from the broker, containing partitions grouped by
+     *                   topic id.
+     * @return Set of {@link TopicPartition} containing topic name and partition id.
+     */
+    private CompletableFuture<SortedSet<TopicPartition>> extractTopicPartitionsFromAssignment(
+            ConsumerGroupHeartbeatResponseData.Assignment assignment) {
+        SortedSet<TopicPartition> assignedPartitions = new TreeSet<>(COMPARATOR);
+
+        List<Uuid> topicsRequiringMetadata = new ArrayList<>();
+        assignment.topicPartitions().forEach(topicPartitions -> {
+            Uuid topicId = topicPartitions.topicId();
+            if (!metadata.topicNames().containsKey(topicId)) {
+                topicsRequiringMetadata.add(topicId);
+            } else {
+                String topicName = metadata.topicNames().get(topicId);
+                topicPartitions.partitions().forEach(tp -> assignedPartitions.add(new TopicPartition(topicName, tp)));
+            }
+
+        });
```

Comment:
I guess it's OK to use `ConsumerMetadata` directly like this as we "own" updating it on the consumer network thread.

Reply:
I think so. Also, the set return by `ConsumerMetadata` is immutable.